### ISSUE TITLE
On token refresh useAuthenticator returns an unauthenticated user object

### ISSUE THREAD

**Issue by jeroenmeulendijks** [aws-amplify/amplify-ui#1683]
### Before creating a new issue, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-ui/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-ui/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md#bug-reports).

### On which framework/platform are you having an issue?

React

### Which UI component?

Authenticator

### How is your app built?

CRA with webpack 5

### Please describe your bug.

This issue might be related to #1333 and/or #1497 

For testing purposes I reduced the token refresh time to 7 minutes, but I assume it will not be different with longer refresh times.

When the token is expired and it should be refreshed and you than navigate to a page which uses useAuthenticator I get the error:
![image](https://user-images.githubusercontent.com/13433769/163276560-85d0c199-2ab3-4da9-9b8e-eed870559dcb.png)

### What's the expected behaviour?

That the token is refreshed by the library and useAuthenticator would still return a valid user object

### Help us reproduce the bug!

- Log in to the application
- Wait until the token needs to be refreshed (I my case 7 minutes)
- Navigate to a page which uses useAuthenticator and uses user.getUserData() and see the error

### Code Snippet

```typescript
// Index.js
ReactDOM.render(
  <Provider store={store}>
    <React.StrictMode>
      <Authenticator.Provider>
        <App />
      </Authenticator.Provider>
    </React.StrictMode>
  </Provider>,
  document.getElementById('root')
);
```

```typescript
// App.js
<Authenticator hideSignUp loginMechanisms={['email']}>
        {() => (
          <main>
            ... Components ...
          </main>
        )}
      </Authenticator>
```

```typescript
// MyComponent.js
const { user } = useAuthenticator((context) => [context.user]);
// Here user cannot be used because its not authenticated
user.getUserData()
```

**Comment by wlee221** [comment]
Hi @jeroenmeulendijks, thanks for raising this issue. We should update `Authenticator` to update user objects whenever token refreshes. We'll take a look into this in our earliest capacity. 

Meanwhile, you should be able to workaround this by using `Auth.currentAuthenticatedUser` whenever token is refreshed. You can use Auth Hub event to know when tokens are refreshed: https://docs.amplify.aws/lib/auth/auth-events/q/platform/js/.

**Comment by jeroenmeulendijks** [comment]
Hi @wlee221, thanks for the quick response.

On the workaround, does that mean I basically need to keep track on my own user object through `Auth.currentAuthenticatedUser` or is there a way in which we somehow can update the user object returned by `useAuthenticator()`.

In the first workaround it basically means we cannot use the `useAuthenticator()` hook at the moment, because we would never know if the returned user object is a valid one. If that's true I guess the best workaround is to implement our own hook which uses the `Hub.listen('auth', ..)` to updates its own user state object. And use that hook instead of the Amplify `useAuthenticator`. Once this issue is resolved we could replace this hook with the Amplify one again.

**Comment by wlee221** [comment]
Looking into this issue now (cc @[ptaranto](https://github.com/ptaranto)). 

> If that's true I guess the best workaround is to implement our own hook which uses the Hub.listen('auth', ..)

Unfortunately, yes -- you'd have to rely on Auth hub events before this gets resolved.

**Comment by wlee221** [comment]
Hi @jeroenmeulendijks, sorry this took a while! The fix #1809 is merged and will be released later today. I'll let you know again after the release.

**Comment by wlee221** [comment]
Update -- this has been released to 

```
@aws-amplify/ui-angular@2.4.1
@aws-amplify/ui-react@2.16.1
@aws-amplify/ui-vue@2.4.1
```

Please let us know if the issue persists 🙏

**Comment by jeroenmeulendijks** [comment]
@wlee221 not sure if I'm doing something wrong, but if I use ui-react@2.16.1 I end up in a never ending loop of user token refreshes when using `user.getUserData()`

I get the user object with:
```const { user } = useAuthenticator((context) => [context.user]);```

**Comment by wlee221** [comment]
Thanks for reporting back, I'll take a look. 

For reproducibility purposes, how and where are you using `user.getUserData()`? Can you add more code snippets around this?

**Comment by wlee221** [comment]
Sorry one more ask:

can you enable a logger like so:

```
import { Logger } from 'aws-amplify';

Logger.LOG_LEVEL = 'VERBOSE';
```

and let us know what you get back when token expires and attempts to refresh itself? Tokens can behave pretty differently based on backend, so just wanted to make sure I have the full context. Thanks for your help!

**Comment by wlee221** [comment]
Hi @jeroenmeulendijks, dived into this a bit yesterday.

I couldn't reproduce your exact infinite loop issue on render, but I found (unrelated?) infinite loop bug that happens when you refresh your app with your refresh token expired. That's been fixed in `@aws-amplify/ui-react@2.17.0`, so hoping that might help. There's a #1868 I'm working that might help as well.

For the actual issue though, I looked at `user.getUserData()` and it will *not* attempt to refresh your token if both your accessToken and idToken are expired. If you're only interested in user attributes or username, I would suggest using `user.attributes` and `user.username` because they do not require token access.

Note that tokens are usually refreshed when you actually make an API call to a AWS service at `Amplify` level (e.g. `Storage.get()`, `Datastore.query`, and even `Auth.currentAuthenticatedUser`) but not necessarily on `user` level. So I would rely on those API calls to get token refreshes, and if you're just wanting to get user attributes I'd use the simpler methods I mentioned above.

**Comment by jeroenmeulendijks** [comment]
Hi @wlee221 ,

I'm a bit lost in where the problem is coming from and what it's related to. When I just upgrade from version 2.15.4 to 2.18.0 I get this behavior when opening a page which does:

```
useEffect(() => {
    Auth.currentAuthenticatedUser().then((user) => {
      user.getUserData(
        (e, data) => {
          if (e) {
            // Assume MFA is disabled
            console.error(e);
            setMfaEnabled(false);
          } else {
            const { PreferredMfaSetting } = data;
            setMfaEnabled(PreferredMfaSetting && PreferredMfaSetting === 'SOFTWARE_TOKEN_MFA');
          }
        },
        { bypassCache: true }
      );
    });
  }, []);
```
Basically what I want to know when this component is loaded, is MFA enabled on this user or not.

I enabled the logging and I see a continuous repeat of:
```
ConsoleLogger.js:170 [DEBUG] 47:24.33 AuthClass - getting current authenticated user
ConsoleLogger.js:170 [DEBUG] 47:24.38 AuthClass - get current authenticated userpool user
ConsoleLogger.js:170 [DEBUG] 47:24.38 AuthClass - get current authenticated userpool user
ConsoleLogger.js:182 [DEBUG] 47:24.38 RestClient - POST https://XXXXXXXXXXXX.appsync-api.eu-west-1.amazonaws.com/graphql
ConsoleLogger.js:182 [DEBUG] 47:24.239 Hub - Dispatching to auth with  {event: 'tokenRefresh', data: undefined, message: 'New token retrieved'}
ConsoleLogger.js:182 [DEBUG] 47:24.239 Hub - Dispatching to auth with  {event: 'tokenRefresh', data: undefined, message: 'New token retrieved'}
ConsoleLogger.js:170 [DEBUG] 47:24.240 AuthClass - getting current authenticated user
ConsoleLogger.js:182 [DEBUG] 47:24.240 Hub - Dispatching to auth with  {event: 'tokenRefresh', data: undefined, message: 'New token retrieved'}
ConsoleLogger.js:170 [DEBUG] 47:24.241 AuthClass - Getting current session
ConsoleLogger.js:182 [DEBUG] 47:24.241 Hub - Dispatching to auth with  {event: 'tokenRefresh', data: undefined, message: 'New token retrieved'}
ConsoleLogger.js:170 [DEBUG] 47:24.241 AuthClass - get current authenticated userpool user
ConsoleLogger.js:182 [DEBUG] 47:24.242 AuthClass - Getting the session from this user: CognitoUser {username: 'j.meulendijks@test.net', pool: CognitoUserPool, Session: null, client: Client, signInUserSession: CognitoUserSession, …}
ConsoleLogger.js:182 [DEBUG] 47:24.242 AuthClass - Succeed to get the user session CognitoUserSession {idToken: CognitoIdToken, refreshToken: CognitoRefreshToken, accessToken: CognitoAccessToken, clockDrift: 0}
```
Above doesn't happen when using 2.15.4 but I cannot understand why. Does this help?

**Comment by jeroenmeulendijks** [comment]
BTW, it seems that it might be related to the configuration of how long a token is valid. On our dev environment the token is only valid for 7 minutes and on prod it's 4 hours. And it seems not reproducable on production and only on our dev setup

**Comment by wlee221** [comment]
Hi @jeroenmeulendijks, I apologize for the late response here. That's interesting that it's only dev-only. I imagine that is due to `StrictMode` on dev, it does a few weird things like [simulating component remounts](https://reactjs.org/docs/strict-mode.html#ensuring-reusable-state) on first render. This could possibly mess up the hub subscription system.

Can you try verify if this is the case? I also merged a related fix in #2183, which will be published soon as well.

**Comment by ErikCH** [comment]
Hi @jeroenmeulendijks ! We haven't heard from you in a bit. I'm going to close this, but feel free to re-open the issue if it is still occurring after using the latest release.
